az-digital/az_quickstart
# Fixes #409, #439 Third Party Libraries and Add Chosen

This Pull request adds Chosen and a method of adding third party client-side libraries to Quickstart. **There is a matching pull request in the scaffolding repository** that adds [Asset Packagist](https://asset-packagist.org/) as a package repository.

## Description
This pull request uses [Asset Packagist](https://asset-packagist.org/) to add chosen and its client side library to Quickstart. This is presently being done by using Asset Packagist, which is a standalone packagist repository that allows importing NPM and Bower assets as if they were composer packages. A topic for discussion is if this provides enough flexibility for client side libraries we may want to import in the future.

It is worth noting that Chosen's library (`harvesthq/chosen`) is **already available** as a composer package. However, it is not configured correctly and lists its package type as `library`, (the default). This means that by default it is installed inside the `vendor` folder rather than where it is needed. This is a long term issue with the chosen library and hasn't been fixed yet as none of the current maintainers have many ties to PHP. There is a pull request regarding this waiting in the chosen repository.

Since the type is not correct, the only way to target it via `composer/installers` would be to mention chosen by name in the composer installer settings in the scaffolding. This didn't seem the best practice for long-term maintainability, so instead I have used an equivalent asset-packagist package, `bower-asset/chosen`.

An equivalent possible solution would be to mention a custom install location for `harvesthq/chosen` in the **scaffolding** repository. I've elected to use the current asset packagist solution in this PR as it addresses the **Third Party Libraries** #439 ticket as well.

This implies that chosen really is being added twice (once by asset-packagist and once from the incorrectly-maintained version that's added as a dependency by chosen), but once is to the Vendor folder where it isn't terribly useful:
```
  - Installing harvesthq/chosen (v1.8.7): Extracting archive
  - Installing drupal/chosen (3.0.1): Extracting archive
  - Installing drupal/chosen_lib (3.0.1)
  - Installing bower-asset/chosen (v1.8.7): Extracting archive
```

## Related Issue
#409 
#439 

## How Has This Been Tested?

- Visit `/admin/config/user-interface/chosen` and verify that it doesn't complain that the chosen library is missing
- Verify that chosen is set to be used only on admin pages
- Verify that chosen is set to always apply on both single select and multiselect.
- Verify that chosen is in use on admin pages.

## Types of changes


## Checklist:
